server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceService.java

```diff
@@ -238,13 +238,19 @@ boolean updateDesiredBalanceAndReroute(DesiredBalanceInput desiredBalanceInput,
             );
         }
 
+        final var unassigned = new HashMap<ShardId, UnassignedInfo>();
+        for (var ignored : routingAllocation.routingNodes().unassigned().ignored()) {
+            assert ignored.unassignedInfo() != null;
+            unassigned.put(ignored.shardId(), ignored.unassignedInfo());
```

Comment:
I think we need something finer-grained than `ShardId` for the key - we might have just created a new copy for an existing shard (e.g. increased `number_of_replicas`). The primary could be assigned, the existing replica could be `DECIDERS_NO` and the new replica could be `NO_ATTEMPT`.

Also I think the only meaningful status here will be `DECIDERS_NO` - the other possibilities can only come from one of the existing-shards allocators I think.